Task 15: Execute queries with joins and relations

yevgenii-v opened the pull request:
and relations

yevgenii-v committed:
Task 15: Execute queries with joins
and relations

SanycoWS commented on app/Http/Resources/BookModelResource.php:
```diff
@@ -0,0 +1,35 @@
+            'id'        => $resource->id,
+            'name'      => $resource->name,
+            'year'      => $resource->year,
+            'category'  => [
```
в таких випадках краще завжди використовувати ресурс. CategoryResource

SanycoWS commented on app/Repositories/Books/BookRepository.php:
```diff
@@ -89,10 +92,58 @@ public function getById(int $id): BookIterator
+            ->first();
+
+        $book = new BookIterator($bookQuery);
+        $book->setCategory(new CategoryIterator((object)[
```
сетери для ітераторів є небезпечними. 
якщо використовуєш сетери, то в тестах постійно потрібно буде перевіряти що в усіх сервісах ти не використовуєш цих сетерів. 
якщо ж книжка залежить від категорії то ти можеш приймати категорію як другий параметр в конструкторі при побудові BookIterator. це як варіант. способів це зробити є багато.

основна роль ітератора це забеспечити наступне:
- мати структуризовану структуру, де ти зможеш за допомогою гетерів отримувати потрібні значення
- забеспечити захист вибраних даних, їх змінити в подальшому коді не буде можливості
- забеспечити однорідну структуру даних, навідмінну від колекцій, до ітератору можна додавати тільки однотипні обєкти
- зменшити використання памяті

yevgenii-v committed:
Task 15: Refactoring
- Setters was removed from iterators.
- Additional iterators and resources was
added.

yevgenii-v committed:
Task 15: Refactoring
